Bugfixes and added output for deployment

tmbgreaves opened the pull request:
This is a general bugfix and verbosity PR, dealing with:

* Parameters not being read from the config file following moving commands from setup.sh to deploy.sh
* An attempt to do sanity checking on the parameters read by jq from the config file, in particular dealing with the case of the config file not being valid json (we could do with a linter, really...)
* Adding in the possibility of docker credentials being supplied in the config file
* Adding in a contact email config entry for use with letsencrypt

closes #34 #31 #26 #25 #15 #27

tmbgreaves committed:
Add more useful output during deployment

tmbgreaves committed:
More useful AKS deploy output

tmbgreaves committed:
Fixing some missing parameter reads, adding comments

tmbgreaves committed:
Improve on error-checking for missing/null parameters

tmbgreaves committed:
Remove errant 'x'

tmbgreaves committed:
Fixing a bit of missing documentation change

tmbgreaves committed:
Remove double-quote

sgibson91 commented:
+1 for linter!!!

tmbgreaves committed:
Adding more deployment output

tmbgreaves committed:
Neatening output around docker password input

tmbgreaves commented:
Added more fixes and documentation -- running this through deployment testing now.

sgibson91 commented:
Here's a CLI JSON linter: https://github.com/zaach/jsonlint

tmbgreaves commented:
Brilliant, thanks @sgibson91

tmbgreaves committed:
Making helm version check more robust

tmbgreaves committed:
Checking version probably isn't fatal

tmbgreaves committed:
Ensure yaml is available for python3 on centos/fedora

tmbgreaves committed:
Allow helm version check to fail

sgibson91 committed:
Merge branch 'master' into master

sgibson91 commented on deploy.sh:
```diff
@@ -88,37 +150,51 @@ if [ -z $SP_APP_ID ] || [ -z $SP_APP_KEY ] || [ -z $SP_TENANT_ID ] ; then
 # Check nodes are ready
-while [[ ! x`kubectl get node | awk '{print $2}' | grep Ready | wc -l` == x${AKS_NODE_COUNT} ]] ; do echo -n $(date) ; echo " : Waiting for all cluster nodes to be ready" ; sleep 15 ; done
+nodecount="$(kubectl get node | awk '{print $2}' | grep Ready | wc -l)"
+while [[ ! x${nodecount} == x${AKS_NODE_COUNT} ]] ; do echo -n $(date) ; echo " : ${nodecount} of ${AKS_NODE_COUNT} nodes ready" ; sleep 15 ; nodecount="$(kubectl get node | awk '{print $2}' | grep Ready | wc -l)" ; done
```
```suggestion
while [[ x${nodecount} -ne x${AKS_NODE_COUNT} ]] ; do echo -n $(date) ; echo " : ${nodecount} of ${AKS_NODE_COUNT} nodes ready" ; sleep 15 ; nodecount="$(kubectl get node | awk '{print $2}' | grep Ready | wc -l)" ; done
```

To fix the never-ending loop on MacOS

tmbgreaves committed:
Fixing mac looping

Big thanks to @sgibson91 for squashing this bug -- defeated me!

Co-Authored-By: sgibson91 <44771837+sgibson91@users.noreply.github.com>

tmbgreaves committed:
Take resource group name from config.json

I'd previously thought that to simplify the config file
we would generate the resource group name from the name
of the binderhub instance, but when deployment happens
from blue button the resource group name is passed in
to the deploy container, so to standardise we should use
resource group name from the config file.

This commit removes a couple of lines that should have
been taken out when the move to config file resource
group naming happened.

tmbgreaves committed:
Merge branch 'master' of git+ssh://github.com/tmbgreaves/binderhub-deploy

sgibson91 committed:
Correct hub_url placement in create_config.py

sgibson91 commented on deploy.sh:
```diff
@@ -46,31 +48,89 @@ else
+    SP_APP_KEY: ${SP_APP_KEY}
+    SP_TENANT_ID: ${SP_TENANT_ID}
+    DOCKER_USERNAME: ${DOCKER_USERNAME}
+    DtOCKER_PASSWORD: ${DOCKER_PASSWORD}
```
```suggestion
    DOCKER_PASSWORD: ${DOCKER_PASSWORD}
```

sgibson91 committed:
Correct syntax for testing if a variable is None

sgibson91 committed:
Add if clauses to catch Docker org when building config file

sgibson91 committed:
Standardising echo output in deploy,sh

tmbgreaves committed:
Fix parameter name typo

Thanks @sgibson91 !

Co-Authored-By: sgibson91 <44771837+sgibson91@users.noreply.github.com>

sgibson91 committed:
Close deploy.sh by printing Binder IP

tmbgreaves committed:
Removing 'x's from previous string comparison method

tmbgreaves committed:
Merge branch 'master' of git+ssh://github.com/tmbgreaves/binderhub-deploy

sgibson91 committed:
Correct if statement to echo message about Docker org

sgibson91 commented on README.md:
```diff
@@ -56,12 +56,18 @@ Fill the quotation marks with your desired namespaces, etc.
   "binderhub": {
     "name": "",          # Name of your BinderHub
     "version": ""        # Helm chart version to deploy, should be 0.2.0-<commit-hash>
+    "contact_email": ""  # Email for letsencrypt https certificate
```
```suggestion
    "contact_email": ""  # Email for letsencrypt https certificate. CANNOT be left blank.
```

This is a hard requirement for running the code.

sgibson91 committed:
Correct syntax error

sgibson91 committed:
Correct syntax

tmbgreaves committed:
Merge pull request #1 from sgibson91/master

Fixes to deploy.sh and create_config.py

tmbgreaves committed:
Update README.md

Co-Authored-By: sgibson91 <44771837+sgibson91@users.noreply.github.com>

tmbgreaves committed:
Update README.md

tmbgreaves committed:
Fix kubectl architecture in Dockerfile

tmbgreaves committed:
Merge branch 'master' of git+ssh://github.com/tmbgreaves/binderhub-deploy

tmbgreaves committed:
Fix cases in template variable naming

tmbgreaves committed:
Fix name for kubernetes input

tmbgreaves committed:
Propagate modified name

tmbgreaves committed:
Fix container name; set container mode

tmbgreaves committed:
Deal with prepending from Azure blue button

tmbgreaves committed:
Use any provided SP for AKS creation

tmbgreaves committed:
Push deployment files and logs to blob storage

tmbgreaves committed:
Push ssh keys

tmbgreaves committed:
Revert to actually checking tiller pod status

tmbgreaves committed:
Use a date-stamped container name for log uploads

tmbgreaves committed:
Fixing azure storage handling bugs

tmbgreaves committed:
Increasing IP address waits

Loops are going through ~30 iterations before finding the
IP address; increasing wait from 5s to 30s and adding an
info note for users

tmbgreaves committed:
Blob storage doesn't like ~ expansion

sgibson91 commented on deploy.sh:
```diff
@@ -168,34 +272,84 @@ python3 $secret_script --apiToken=$apiToken \
 echo "--> Updating Helm chart"
-helm upgrade $BINDERHUB_NAME jupyterhub/binderhub \
+helm upgrade $HELM_BINDERHUB_NAME jupyterhub/binderhub \
 --version=$BINDERHUB_VERSION \
```
```suggestion
--version=$HELM_BINDERHUB_VERSION \
```
For consistency? Just a suggestion, feel free to ignore.

sgibson91 commented on deploy.sh:
```diff
@@ -84,62 +147,99 @@ AKS_NAME=`echo ${BINDERHUB_NAME} | tr -cd '[:alnum:]-' | cut -c 1-59`-AKS
   fi
 else
-  echo "Attempting to log in to Azure with service principal"
+  echo "--> Attempting to log in to Azure with service principal ${SP_APP_ID}"
```
I'm not 100% sure we should be echoing the SP appID here. It's not really the kind of thing you want lingering in logs and/or terminal history.

sgibson91 replied:
I'd be happier with:
```
SP_NAME=`az ad sp show --id ${SP_APP_ID} --query displayName -o tsv`
```
and echoing `SP_NAME`.

tmbgreaves replied:
Fixed! https://github.com/alan-turing-institute/binderhub-deploy/pull/22/commits/64a5d159f06e8628eb95b6cccd80a619823f7361

sgibson91 commented on README.md:
```diff
@@ -117,7 +123,7 @@ The user should check the [Azure Portal](https://portal.azure.com/#home) to veri
 To deploy [Binderhub](https://binderhub.readthedocs.io/) to Azure use the deploy button below.
 
-[![Deploy to Azure](https://azuredeploy.net/deploybutton.svg)](https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2Ftmbgreaves%2Fbinderhub-deploy%2Fchange-urls-to-upstream%2Fazure%2Fpaas%2Farm%2Fazure.deploy.json)
+[![Deploy to Azure](https://azuredeploy.net/deploybutton.svg)](https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2Ftmbgreaves%2Fbinderhub-deploy%2Fmaster%2Fazure%2Fpaas%2Farm%2Fazure.deploy.json)
```
```suggestion
[![Deploy to Azure](https://azuredeploy.net/deploybutton.svg)](https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2Falan-turing-institute%2Fbinderhub-deploy%2Fmaster%2Fazure%2Fpaas%2Farm%2Fazure.deploy.json)
```

Update the button link to point to the parent repo. This should be the last thing we commit before merging.

tmbgreaves committed:
Make deployment output blob name static

tmbgreaves committed:
Update README.md

tmbgreaves committed:
Update README.md

tmbgreaves committed:
Update README.md

tmbgreaves committed:
Update README.md

tmbgreaves committed:
Update README.md

tmbgreaves committed:
Update README.md

sgibson91 commented on README.md:
```diff
@@ -117,7 +123,38 @@ The user should check the [Azure Portal](https://portal.azure.com/#home) to veri
+
+To monitor the progress of a blue-button deployment, go to the [Azure portal](https://portal.azure.com/) and select 'Resource Groups' from the left hand pane. Then in the central pane select the resource group you chose to deploy into. This will give you a right hand pane containing the resources within the group. You may need to 'refresh' until you see a new container instance. When it appears,  ...
+
+### Retrieving deployment output from Azure
```
We may want to add a note in this section about pulling over the ssh keys. I think they're automatically generated as `id_rsa`, so we'll want to remind people to rename them when they're copying them over so they don't overwrite they're own `id_rsa` keys.

tmbgreaves replied:
Good thought! I can fix that in the upload.

tmbgreaves replied:
Fixed in https://github.com/alan-turing-institute/binderhub-deploy/pull/22/commits/ebaf54581f86889c4a60a359022b551d93a28637

sgibson91 committed:
Add image retrieving sub ID

sgibson91 committed:
Include sub ID image in README

sgibson91 committed:
Add create-for-rbac image

sgibson91 committed:
Include create for rbac image in README

sgibson91 committed:
Remove unrequired image

tmbgreaves committed:
Make storage account unique with timestamp

tmbgreaves committed:
Merge branch 'master' of git+ssh://github.com/tmbgreaves/binderhub-deploy

tmbgreaves committed:
Update deploy.sh

Co-Authored-By: sgibson91 <44771837+sgibson91@users.noreply.github.com>

tmbgreaves committed:
Change SP ID to name for log sanity

tmbgreaves committed:
Merge branch 'master' of git+ssh://github.com/tmbgreaves/binderhub-deploy

tmbgreaves committed:
Tag ssh key with BinderHub name

tmbgreaves committed:
Single definition of CONTAINER_NAME

tmbgreaves committed:
Merge pull request #2 from sgibson91/docs-update

Updates docs for Service Principal.

sgibson91 commented:
For the button, there seems to be a bug with logging into azure. Shall we just merge this and start with a new PR to debug that problem?

tmbgreaves committed:
Don't do az commands before logging in!

tmbgreaves committed:
Merge branch 'master' of git+ssh://github.com/tmbgreaves/binderhub-deploy

sgibson91 commented on deploy.sh:
```diff
@@ -168,34 +272,86 @@ python3 $secret_script --apiToken=$apiToken \
+  --jupyterhub_ip=$jupyterhub_ip \
+  --force
+fi
 
```
```suggestion

helm repo update
```

Try to avoid helm error in button deployment.

sgibson91 commented:
Some error outputs from latest attempt of button deployment:

```
--> Updating Helm chart
Error: failed to download "jupyterhub/binderhub" (hint: running `helm repo update` may help)
```
I don't really understand why this has cropped. I can't see any spelling mistakes/syntax errors. Maybe we should just run `helm repo update` again before `helm upgrade`. Inline comment left.

```
--> Creating storage account
ERROR: hubtest-2019-05-31-09-13 is not a valid storage account name. Storage account name must be between 3 and 24 characters in length and use numbers and lower-case letters only.
```
I don't think we're allowed hyphens in the storage account name.

```
--> Creating storage container: 
ERROR: az storage container create: error: argument --name/-n: expected one argument
```
Looks like a parsing issue of `CONTAINER_NAME`, probably related to above.

```
--> Pushing log files
ERROR: az storage blob upload-batch: error: argument --destination/-d: expected one argument
```
Similar issue here.

It also looks like the logs get trimmed from the top as well, which is not ideal so worth warning about.
![azure_container_logs](https://user-images.githubusercontent.com/44771837/58701396-b1a39200-839a-11e9-87c0-91ad2f33db6b.png)

sgibson91 commented on deploy.sh:
```diff
@@ -168,34 +272,86 @@ python3 $secret_script --apiToken=$apiToken \
+  # Create a storage account
+  echo "--> Creating storage account"
+  CONTAINER_NAME="${CONTAINER_NAME}"
+  STORAGE_ACCOUNT_NAME="${BINDERHUB_NAME}-$(date +'%Y-%m-%d-%H-%M')"
```
```suggestion
  STORAGE_ACCOUNT_NAME="${BINDERHUB_NAME}$(date +'%Y%m%d%H%M')"
```
Would this work?

tmbgreaves committed:
Update deploy.sh

Co-Authored-By: sgibson91 <44771837+sgibson91@users.noreply.github.com>

tmbgreaves committed:
Update deploy.sh

Good fix, thank you @sgibson91 !

Co-Authored-By: sgibson91 <44771837+sgibson91@users.noreply.github.com>

tmbgreaves commented:
That's a nuisance with the trimmed logs; hopefully all the really useful output gets sent to log files which end up in the storage container, and if there are any errors I guess the most useful bit of the 'main' log is the last bit.

sgibson91 commented:
Yes, hopefully the last bit of the log is either the successful deployment or the last error. I'll give the button another go, then I'm happy to merge this and work on the Binder Internal Server Errors in a new PR 👍

sgibson91 commented:
```
--> Waiting 5 seconds before attempting helm version check again
Client: &version.Version{SemVer:"v2.14.0", GitCommit:"05811b84a3f93603dd6c2fcfe57944dfa7ab7fd0", GitTreeState:"clean"}
Error: could not find a ready tiller pod
--> helm version attempt 3 of 3 failed
--> Please check helm versions manually later
```
Might have to extend tiller waiting time to 30 seconds.

```
--> Retrieving JupyterHub IP
Error from server (NotFound): namespaces "hubtest" not found
Sleeping 30s before checking again
```
Seems to be stuck in a loop checking for the IP address

sgibson91 commented on deploy.sh:
```diff
@@ -84,62 +147,99 @@ AKS_NAME=`echo ${BINDERHUB_NAME} | tr -cd '[:alnum:]-' | cut -c 1-59`-AKS
+
+# Waiting until tiller pod is ready
+tillerStatus="$(kubectl get pods --namespace kube-system | grep ^tiller | awk '{print $3}')"
+while [[ ! x${tillerStatus} == xRunning ]] ; do echo -n $(date) ; echo " : tiller pod status : ${tillerStatus} " ; sleep 5 ; tillerStatus="$(kubectl get pods --namespace kube-system | grep ^tiller | awk '{print $3}')" ; done
```
```suggestion
while [[ ! x${tillerStatus} == xRunning ]] ; do echo -n $(date) ; echo " : tiller pod status : ${tillerStatus} " ; sleep 30 ; tillerStatus="$(kubectl get pods --namespace kube-system | grep ^tiller | awk '{print $3}')" ; done
```
Wait 30 seconds for a ready tiller pod, instead of 5.

tmbgreaves committed:
Update deploy.sh

Co-Authored-By: sgibson91 <44771837+sgibson91@users.noreply.github.com>

sgibson91 commented:
This message cropped up again:
```
--> Updating Helm chart
Error: failed to download "jupyterhub/binderhub" (hint: running `helm repo update` may help)
```
I don't like it 👎

sgibson91 commented on deploy.sh:
```diff
@@ -168,34 +272,87 @@ python3 $secret_script --apiToken=$apiToken \
+  # 
+  # Create a storage account
+  echo "--> Creating storage account"
+  CONTAINER_NAME="${CONTAINER_NAME}"
```
@tmbgreaves Where does this variable get set? I think this is why the pushing of the logs is still failing.

sgibson91 replied:
```suggestion
  CONTAINER_NAME="${HELM_BINDERHUB_NAME}setuplogs"
```
Might work?

sgibson91 commented:
> This message cropped up again:
> 
> ```
> --> Updating Helm chart
> Error: failed to download "jupyterhub/binderhub" (hint: running `helm repo update` may help)
> ```
> 
> I don't like it 👎

I think this is due to inconsistent variable names throughout the script. Some are `BINDERHUB_NAME`/`BINDERHUB_VERSION` whereas others are `HELM_BINDERHUB_NAME`/`HELM_BINDERHUB_VERSION`.

tmbgreaves committed:
Removing a rogue variable change

I'd failed to catch a rogue batch replacement which
generated a nonexistant variable.

tmbgreaves committed:
Fixing a beautifully pointless (un)definition

Having tried to decide whether to hardcode the
container name into commands or define it once,
I managed to commit a version of my code which
did neither.

Oops.

tmbgreaves committed:
Length checks on container account name

Container account name can't be more than 24 characters; with
the long date string this was easy to break. This commit changes
datestamp to short random string to leave more space for the
config-supplied name, and ensures string doesn't go over the
24-character limit.

tmbgreaves committed:
Fix variable name checks

Allowing - and . in helm name, and only allow alphanumerics in
the storage account name.

tmbgreaves committed:
Update blob interaction documentation

tmbgreaves committed:
Change to upstream repo for blue button

Co-Authored-By: sgibson91 <44771837+sgibson91@users.noreply.github.com>

sgibson91 reviewed (approved):
🎉
